src/modules/navigator/MissionFeasibility/FeasibilityChecker.cpp

```diff
@@ -0,0 +1,692 @@
+#include <systemlib/mavlink_log.h>
+#include <px4_platform_common/events.h>
+#include <drivers/drv_pwm_output.h>
+#include "../mission_block.h"
+#include <lib/mathlib/mathlib.h>
+#include <lib/geo/geo.h>
+
+FeasibilityChecker::FeasibilityChecker() :
+	ModuleParams(nullptr)
+{
+}
+
+void FeasibilityChecker::reset()
+{
+	_mission_validity_failed = false;
+	_takeoff_failed = false;
+	_fixed_wing_landing_failed = false;
+	_distance_first_waypoint_failed = false;
+	_distance_between_waypoints_failed = false;
+	_below_home_alt_failed = false;
+	_fixed_wing_land_approach_failed = false;
+	_takeoff_land_available_failed = false;
+
+	_found_item_with_position = false;
+	_has_vtol_takeoff = false;
+	_has_takeoff = false;
+
+	_landing_valid = false;
+	_do_land_start_index = -1;
+	_landing_approach_index = -1;
+	_mission_item_previous = {};
+
+	_first_waypoint_found = false;
+
+	_last_lat = NAN;
+	_last_lon = NAN;
+	_last_cmd = -1;
+}
+
+void FeasibilityChecker::updateData()
+{
+	home_position_s home = {};
+
+	if (_home_pos_sub.copy(&home)) {
+		if (home.valid_hpos) {
+			_home_lat_lon = matrix::Vector2d(home.lat, home.lon);
+		}
+
+		if (home.valid_alt) {
+			_home_alt_msl = home.alt;
+		}
+	}
+
+	vehicle_status_s status = {};
+
+	if (_status_sub.copy(&status)) {
+		_vehicle_type = status.vehicle_type;
+		_is_vtol = status.is_vtol;
+	}
+
+	vehicle_land_detected_s land_detected = {};
+
+	if (_land_detector_sub.copy(&land_detected)) {
+		_is_landed = land_detected.landed;
+	}
+
+}
+
+void FeasibilityChecker::processNextItem(mission_item_s &mission_item, const int current_index, const int total_count)
+{
+	if (current_index == 0) {
+		reset();
+		updateData();
+	}
+
+	doCommonChecks(mission_item, current_index);
+
+	if (_is_vtol) {
+		doVtolChecks(mission_item, current_index);
+
+	} else if (_vehicle_type == vehicle_status_s::VEHICLE_TYPE_FIXED_WING) {
+		doFixedWingChecks(mission_item, current_index);
+
+	} else {
+		doMulticopterChecks(mission_item, current_index);
+	}
+
+	if (current_index == total_count) {
+		_takeoff_land_available_failed = !checkTakeoffLandAvailable();
+	}
+
+	_mission_item_previous = mission_item;
+}
+
+void FeasibilityChecker::doCommonChecks(mission_item_s &mission_item, const int current_index)
+{
+	if (!_mission_validity_failed) {
+		_mission_validity_failed = !checkMissionItemValidity(mission_item, current_index);
+	}
+
+	if (!_distance_between_waypoints_failed) {
+		_distance_between_waypoints_failed = !checkDistancesBetweenWaypoints(mission_item);
+	}
+
+	if (!_distance_first_waypoint_failed) {
+		_distance_first_waypoint_failed = !checkDistanceToFirstWaypoint(mission_item);
+	}
+
+	if (!_below_home_alt_failed) {
+		_below_home_alt_failed = !checkIfBelowHomeAltitude(mission_item, current_index);
+	}
+
+	if (!_takeoff_failed) {
+		_takeoff_failed = !checkTakeoff(mission_item);
+	}
+}
+
+void FeasibilityChecker::doVtolChecks(mission_item_s &mission_item, const int current_index)
+{
+	if (!_fixed_wing_landing_failed) {
```

Comment:
checkFixedWingLanding is can also be used for VTOL?

Reply:
@sfuhrer It's just a bad name for it, but yes, it's the common part of the checks for FW and VTOL.

Reply:
addressed